docs/design/proposal-troubleshoot-consolidated-cli.md

```diff
@@ -0,0 +1,249 @@
+  - returns any errors
+- `AnalyzeBundle(spec{}, bundlePath string) (results{},error)`
+  - run analysers from the spec and return the analysis results struct
+  - returns analysis struct and errors
+- `ArchiveBundle(bundlePath string, compressionMethod string, destination string) error`
+  - generates a tar archive of the bundle directory at the specified path with optional compression
+  - takes bundle path, compression method and destination as parameters.
+  - returns errors
+- `ParseSpecFile(specFile string) (spec{},error)`
+  - parses a spec file and returns an unmarshalled troubleshoot spec struct.
+  - could be used for linting
+- `ServeBundle(bundlePath string, kubeconfig string, port int)`
+  - starts a sbctl server using the specified bundle and port, outputting a kubeconfig at a specified location.
+
+### Usage patterns
+
+- generate a support bundle
+
+  `troubleshoot collect supportbundle.yaml`
+
+  `troubleshoot collect supportbundle.yaml secrets/default/kotsadm-appslug-supportbundle`
+
+  `troubleshoot collect https://kots.io`
+
+  in the troubleshoot cli code this could look like:
+
+```
+spec,_ := troubleshoot.ParseSpecFile("spec.yaml")
+bundleFile,_ := troubleshoot.CollectBundle(spec)
+analysisResults,_ := troubleshoot.AnalyzeBundle(spec,bundleFile)
+```
+
+- use a spec to return a go/no-go preflight outcome
+
+  `troubleshoot preflight spec.yaml`
+
+  This should not only clearly state any reasons for failing, but also use standardised exit codes that can be used by automation tools.
+
+  We should also continue to accept standard in, accepting any preflights from a yaml multidoc.
+
+  `helm template | troubleshoot preflight -`
+
+- Interact with an existing support bundle with kubectl.
+
+  `troubleshoot inspect support-bundle-12-12-2001.tar.gz`
+
+  This should set the `KUBECONFIG` environment variable, as well as a `TROUBLESHOOT_SHELL` environment variable before spawning a subshell.
+  This is to allow prompts and shell environments to be able to detect that they're running in a troubleshoot spawned shell, much like tmux et-al.
+
+  `troubleshoot inspect --non-interactive support-bundle-12-12-2001.tar.gz`
+
+  this should behave much like the `sbctl serve` command does today, printing the path to it's kubeconfig location to stdout or optionally take a flag to specify the kubeconfig location.
+
+  `troubleshoot inspect --non-interactive -f .kube/config support-bundle-12-12-2001.tar.gz`
+
+- Redact an existing spec
+
+  `troubleshoot redact -f redactors.yaml support-bundle.tar.gz`
+
+- Re-run analysers against an existing support bundle
+
+  `troubleshoot analyze -f spec.yaml support-bundle.tar.gz`
+
+### Example help text
+
+Overall top level command:
+
+```
+Kubernetes diagnostics assistant
+
+Usage:
+  troubleshoot [command]
+
+Available Commands:
+  help        This help text
+  collect     Run collectors, redactors and analyzers, store the result
```

Comment:
a thought: should we always run analyzers AOT (ahead of time) and store the results...? or should we JIT execute analyzers whenever you load up a bundle? as that would allow running against the latest analyzers on old bundles etc.

obviously you can still use `analyze` to re-run against an existing bundle, but does it make sense to run it upfront at all every time...?

we could technically still have a `troubleshoot support-bundle` command that ran collectors, redactors and analyzers and gave you the interactive UI showing the state of analysis still in a single subcommand...

Reply:
so to maintain functionality for vendor portal we'd still need to run analyzers up front to generate the analasys.json.

it's worth thinking about the impact that JIT analasys would have on vendor portal